rtloader/three/three.cpp

```diff
@@ -483,70 +483,86 @@ PyObject *Three::_importFrom(const char *module, const char *name)
 
 PyObject *Three::_findSubclassOf(PyObject *base, PyObject *module)
 {
+    // baseClass is not a Class type
     if (base == NULL || !PyType_Check(base)) {
         setError("base class is not of type 'Class'");
         return NULL;
     }
 
+    // module is not a Module object
     if (module == NULL || !PyModule_Check(module)) {
         setError("module is not of type 'Module'");
         return NULL;
     }
 
     PyObject *dir = PyObject_Dir(module);
     if (dir == NULL) {
+        PyErr_Clear();
         setError("there was an error calling dir() on module object");
         return NULL;
     }
 
     PyObject *klass = NULL;
     for (int i = 0; i < PyList_GET_SIZE(dir); i++) {
-        // get symbol name
-        char *symbol_name;
-        PyObject *symbol = PyList_GetItem(dir, i);
-        if (symbol != NULL) {
-            symbol_name = as_string(symbol);
-            if (symbol_name == NULL)
-                continue;
-        } else {
-            // Gets exception reason
-            PyObject *reason = PyUnicodeDecodeError_GetReason(PyExc_IndexError);
+        // Reset `klass` at every iteration so its state is always clean when we
+        // continue the loop or return early. Reset at first iteration is useless
+        // but it keeps the code readable.
+        Py_XDECREF(klass);
+        klass = NULL;
 
-            // Clears exception and sets error
-            PyException_SetTraceback(PyExc_IndexError, Py_None);
-            setError((const char *)PyBytes_AsString(reason));
+        // get the symbol in current list item
+        PyObject *symbol = PyList_GetItem(dir, i);
+        if (symbol == NULL) {
+            // This should never happen as it means we're out of bounds
+            PyErr_Clear();
+            setError("there was an error browsing dir() output");
             goto done;
         }
 
+        // get symbol name, as_string returns a new object that we have to free
+        char *symbol_name = as_string(symbol);
+        if (symbol_name == NULL) {
+            // as_string returns NULL if `symbol` is not a string object
+            // and raises TypeError. Let's clear the error and keep going.
+            PyErr_Clear();
+            continue;
+        }
+
+        // get symbol instance. It's a new ref but in case of success we don't
+        // DecRef since we return it and the caller will be owner
         klass = PyObject_GetAttrString(module, symbol_name);
         ::free(symbol_name);
         if (klass == NULL) {
+            PyErr_Clear();
```

Comment:
I think it's not required due to `as_string` implementation, but it doesn't harm and better mimics the `two` code.